### ISSUE TITLE
Add ability to control scheme generation modes

### ISSUE THREAD

**Issue by kwridan** [tuist/tuist#458]
## Need/problem

In a large project, generating the entire workspace can yield several schemes, one for each target. Often times projects contain one primary target and a few supporting targets such as a test targets. 

Tuist has the ability to generate custom schemes which allows developers to specify a scheme to group all targets, however currently this is additive (i.e. Tuist default scheme per target + custom schemes). 

To reduce the noise this may generate it would be great to have a mechanism to control the behavior of those default schemes.

### Changing schemes type

```swift
let project = Project(name: "App"
                      targets: [ 
                        // ...
                      ]
                      schemes: .defaults)

let project = Project(name: "App"
                      targets: [ 
                        // ...
                      ]
                      schemes: .defaults(with: [schemeA, schemeB]))

let project = Project(name: "App"
                      targets: [ 
                        // ...
                      ]
                      schemes: .custom([schemeA, schemeB]))

// Possibly restore the project scheme from #55 
// That includes all targets in one scheme
let project = Project(name: "App"
                      targets: [ 
                        // ...
                      ]
                      schemes: .combinedScheme)
```

### Adding a separate configuration option

```swift
let config = TuistConfig(generationOptions: .options(
    schemes: .defaultsWithCustom // there must be a better name for this one :p 
))

let config = TuistConfig(generationOptions: .options(
    schemes: .customOnly
))

let config = TuistConfig(generationOptions: .options(
    schemes: .none
))

// Possibly this option, though if a project manifest specifies schemes 
// should this be additive? or instead of?
let config = TuistConfig(generationOptions: .options(
    schemes: .combined
))

```

Still needs some more thought on how how best to achieve this and which options are most useful.

# Alternatives

Manually specify each of the target un-desired schemes, mark them as non-shared and add `xcuserdata` to `.gitignore`

e.g.
```swift
let testScheme = Scheme(name: "AppTests", shared: false)
```

Sadly those schemes will still be visible in Xcode 

# Previous use cases

- [#55: Project wide scheme](https://github.com/tuist/tuist/issues/55)
- [#97: Auto-create schemes](https://github.com/tuist/tuist/issues/97)
- [#199: Support defining custom schemes](https://github.com/tuist/tuist/issues/199)
- [#289: Support ignoring schemes](https://github.com/tuist/tuist/issues/289)

**Comment by pepibumur** [comment]
What about something along the lines of:

```swift
let config = TuistConfig(generationOptions: [
  .schemes([
    .all: .none,
    .project("FrameworkX"): .custom,
    .project("MyApp"): .custom
  ])
])
```

That way the schemes generation could be configured in a per-project basis. I think the type that we create to indicate which projects configuration applies can be reused by other configuration options:

```swift
enum TuistConfigurationProject: Hashable {
  case all
  case project(String)
}
```

**Comment by sgrgrsn** [comment]
Would be really nice to get something like this! I don't really have an opinion about how it should be specified, but I think it would be nice if Tuist will create a combined scheme for frameworks, so we don't have a FrameworkScheme and FrameworkTestsScheme for each and every framework in the workspace.

**Comment by sgrgrsn** [comment]
I would like to contribute on this one, but I have only been playing around with Tuist for now. Do you have any suggestion on how I could get started contributing to this feature?

**Comment by kwridan** [comment]
Hi @sgrgrsn! 

I'm glad to hear you're interested in contributing to Tuist!

@pepibumur has put together a fantastic contributors guide to help them get started on which you can find on https://docs.tuist.io/contribution-tuist

Once you've got Tuist up and running locally you can have browse through the source code. There have been a few PRs to update schemes recently, those should help show cases the components that would need modifications to support such feature:

- #529:  Defining Archive actions on schemes 
- #728: TestAction.targets testable
- #730: Making scheme generator more generic

The primary component is the `SchemeGenerator` and the associated model & manifest types.

Feel free to ask questions either here or on Slack (https://slack.tuist.io).

**Comment by sgrgrsn** [comment]
Thanks - I'll have a look at it as soon as possible 👍

**Comment by danyf90** [comment]
Implemented in 3.0
